cli/command/completion/functions.go

```diff
@@ -0,0 +1,94 @@
+package completion
+
+import (
+	"os"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/command/formatter"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/spf13/cobra"
+)
+
+// ValidArgsFn a function to be used by cobra command as `ValidArgsFunction` to offer command line completion
+type ValidArgsFn func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective)
+
+// CompleteImageNames offers completion for images present within the local store
+func CompleteImageNames(dockerCli command.Cli) ValidArgsFn {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		list, err := dockerCli.Client().ImageList(cmd.Context(), types.ImageListOptions{})
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+		var names []string
+		for _, image := range list {
+			names = append(names, image.RepoTags...)
+		}
+		return names, cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+// CompleteContainerNames offers completion for container names and IDs
+// By default, only names are returned.
+// Set DOCKER_COMPLETION_SHOW_CONTAINER_IDS=yes to also complete IDs.
+func CompleteContainerNames(dockerCli command.Cli, all bool, filters ...func(types.Container) bool) ValidArgsFn {
```

Comment:
I'm ok with keeping this for now, and iterate on this, but somewhat on the fence on the `bool` argument (which could also be a "filter" / "option") and  doing the filtering client-side

I _think_ most (if not all) of the filtering we need can be done with a `--filter` (although in some cases, we'd have to pass multiple filters). For example;

```bash
docker ps --filter status=exited --filter status=created
CONTAINER ID   IMAGE     COMMAND   CREATED          STATUS                      PORTS     NAMES
7b1647808d7b   busybox   "sh"      56 minutes ago   Exited (0) 56 minutes ago             stopped-container
2ee1ae95ecf4   busybox   "sh"      56 minutes ago   Created                               created-container
```

We could still look at implementing `!=` as filter (`--filter status!=running`), but in most cases, we'd still have to be more specific (e.g. a `paused` container cannot be started, so only excluding `running` likely won't work)